File: typescript/tests/metagrammar/escapeSequences.test.ts
import { MetaGrammar, Parser } from '../../src/index.js';

describe('MetaGrammar - Escape Sequences', () => {
  test('newline escape', () => {
    const grammar = String.raw`
      Newline <- "\n";
    `;

    const rules = MetaGrammar.parseGrammar(grammar);
    const parser = new Parser({ topRuleName: 'Newline', rules, input: '\n' });
    const parseResult = parser.parse();
    const result = parseResult.root;
    expect(result).not.toBeNull();
    expect(result.len).toBe(1);
  });

  test('tab escape', () => {
    const grammar = String.raw`
      Tab <- "\t";
    `;

    const rules = MetaGrammar.parseGrammar(grammar);
    const parser = new Parser({ topRuleName: 'Tab', rules, input: '\t' });
    const parseResult = parser.parse();
    const result = parseResult.root;
    expect(result).not.toBeNull();
    expect(result.len).toBe(1);
  });

  test('backslash escape', () => {
    const grammar = String.raw`
      Backslash <- "\\";
    `;

    const rules = MetaGrammar.parseGrammar(grammar);
    const parser = new Parser({ topRuleName: 'Backslash', rules, input: '\\' });
    const parseResult = parser.parse();
    const result = parseResult.root;
    expect(result).not.toBeNull();
    expect(result.len).toBe(1);
  });

  test('quote escapes', () => {
    const grammar = String.raw`
      DoubleQuote <- "\"";
      SingleQuote <- '\'';
    `;

    const rules = MetaGrammar.parseGrammar(grammar);

    let parser = new Parser({ topRuleName: 'DoubleQuote', rules, input: '"' });
    let parseResult = parser.parse();
    let result = parseResult.root;
    expect(result).not.toBeNull();

    parser = new Parser({ topRuleName: 'SingleQuote', rules, input: "'" });
    parseResult = parser.parse();
    result = parseResult.root;
    expect(result).not.toBeNull();
  });

  test('escaped sequence in string', () => {
    const grammar = String.raw`
      Message <- "Hello\nWorld";
    `;

    const rules = MetaGrammar.parseGrammar(grammar);
    const parser = new Parser({ topRuleName: 'Message', rules, input: 'Hello\nWorld' });
    const parseResult = parser.parse();
    const result = parseResult.root;
    expect(result).not.toBeNull();
    expect(result.len).toBe(11);
  });
});

File: dart/test/metagrammar/escape_sequences_test.dart
import 'package:squirrel_parser/squirrel_parser.dart';
import 'package:test/test.dart';

void main() {
  group('MetaGrammar - Escape Sequences', () {
    test('newline escape', () {
      const grammar = r'''
        Newline <- "\n";
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);
      final parser = Parser(topRuleName: 'Newline', rules: rules, input: '\n');
      var parseResult = parser.parse();
      final result = parseResult.root;
      expect(result, isNotNull);
      expect(result.len, equals(1));
    });

    test('tab escape', () {
      const grammar = r'''
        Tab <- "\t";
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);
      final parser = Parser(topRuleName: 'Tab', rules: rules, input: '\t');
      var parseResult = parser.parse();
      final result = parseResult.root;
      expect(result, isNotNull);
      expect(result.len, equals(1));
    });

    test('backslash escape', () {
      const grammar = r'''
        Backslash <- "\\";
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);
      final parser = Parser(topRuleName: 'Backslash', rules: rules, input: '\\');
      var parseResult = parser.parse();
      final result = parseResult.root;
      expect(result, isNotNull);
      expect(result.len, equals(1));
    });

    test('quote escapes', () {
      const grammar = r'''
        DoubleQuote <- "\"";
        SingleQuote <- '\'';
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);

      var parser = Parser(topRuleName: 'DoubleQuote', rules: rules, input: '"');
      var parseResult = parser.parse();
      var result = parseResult.root;
      expect(result, isNotNull);

      parser = Parser(topRuleName: 'SingleQuote', rules: rules, input: "'");
      parseResult = parser.parse();
      result = parseResult.root;
      expect(result, isNotNull);
    });

    test('escaped sequence in string', () {
      const grammar = r'''
        Message <- "Hello\nWorld";
      ''';

      final rules = MetaGrammar.parseGrammar(grammar);
      final parser = Parser(topRuleName: 'Message', rules: rules, input: 'Hello\nWorld');
      var parseResult = parser.parse();
      final result = parseResult.root;
      expect(result, isNotNull);
      expect(result.len, equals(11));
    });
  });
}
